### PR TITLE
JCC writes build errors to a file for reliable future access.

### PR DESCRIPTION
**Pull request by Alan32Liu**
Some projects (e.g., `mosh`) do not report the root build error sent to `stderr`.
Saving the errors to a file so that we won't miss any and can always access them later.

Reproduce:
```bash
# In OSS-Fuzz repo:
PROJECT=mosh; \
echo -e "\nENV CC=/usr/local/bin/clang-jcc\nENV CXX=/usr/local/bin/clang++-jcc\nENV FUZZING_LANGUAGE=c++\nENV JCC_GENERATE_AST_DIR=/workspace/out/$PROJECT" >> "projects/$PROJECT/Dockerfile"; \
python infra/helper.py build_image "$PROJECT" --pull && \
docker run -ti gcr.io/oss-fuzz/$PROJECT /bin/bash
```

```bash
# In docker container:
wget https://storage.googleapis.com/clusterfuzz-builds/jcc/clang-jcc -O /usr/local/bin/clang-jcc && wget https://storage.googleapis.com/clusterfuzz-builds/jcc/clang++-jcc -O /usr/local/bin/clang++-jcc && compile
```

With this change, we can see errors like:
```
conftest.cpp:12:10: fatal error: 'ac_nonexistent.h' file not found
#include <ac_nonexistent.h>
         ^~~~~~~~~~~~~~~~~~
1 error generated.
/usr/bin/ld: cannot find -lutempter
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
conftest.cpp:76:10: fatal error: 'libutil.h' file not found
#include <libutil.h>
         ^~~~~~~~~~~
1 error generated.
conftest.cpp:78:10: fatal error: 'sys/endian.h' file not found
#include <sys/endian.h>
         ^~~~~~~~~~~~~~
1 error generated.
/usr/bin/ld: /tmp/conftest-18d81d.o: in function `main':
conftest.cpp:(.text+0x10): undefined reference to `pledge'
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
conftest.cpp:89:10: fatal error: 'CommonCrypto/CommonCrypto.h' file not found
#include <CommonCrypto/CommonCrypto.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
/usr/bin/ld: /tmp/conftest-552a22.o: in function `main':
conftest.cpp:(.text+0x10): undefined reference to `forkpty'
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
conftest.cpp:60:10: fatal error: 'tr1/memory' file not found
#include <tr1/memory>
         ^~~~~~~~~~~~
1 error generated.
conftest.cpp:78:10: error: builtin functions must be directly called
  (void) __builtin_bswap64;
         ^
1 error generated.
```

### CODE REVIEW

File: infra/base-images/base-builder/jcc/jcc.go [hunk 1]
```diff
@@ -341,6 +341,18 @@ func CppifyHeaderIncludes(contents string) (string, error) {
 	return contents, nil
 }
 
+func appendStringToFile(filepath, new_content string) error {
+    // Appends |new_content| to the content of |filepath|.
```
Comment (jonathanmetzman) [inline infra/base-images/base-builder/jcc/jcc.go:345]: formatting seems off. can you run go format.

File: infra/base-images/base-builder/jcc/jcc.go [hunk 2]
```diff
@@ -391,6 +403,7 @@ func main() {
 		//  to make build systems happy.
 		fmt.Print(out)
 		fmt.Fprint(os.Stderr, errstr)
+               appendStringToFile("/out/err.log", errstr)
```
Comment (jonathanmetzman) [inline infra/base-images/base-builder/jcc/jcc.go:406]: You may want to do this everywhere there is an Fprintf(os.stderr).

File: infra/base-images/base-builder/jcc/jcc.go (hunk 1 above)
Comment (Alan32Liu) [reply infra/base-images/base-builder/jcc/jcc.go]: Done!

File: infra/base-images/base-builder/jcc/jcc.go (hunk 2 above)
Comment (Alan32Liu) [reply infra/base-images/base-builder/jcc/jcc.go]: Fixed, thanks!

File: infra/base-images/base-builder/jcc/jcc.go [hunk 3]
```diff
@@ -341,6 +341,24 @@ func CppifyHeaderIncludes(contents string) (string, error) {
...
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	_, err = file.WriteString(new_content)
+	return err
+}
+
+func RecordError(errstr string) {
```
Comment (jonathanmetzman) [inline infra/base-images/base-builder/jcc/jcc.go:356]: nit: Maybe change this to WriteStdErrOut and take both as input, write both to stdout/stderr and record stderr.

**Review by jonathanmetzman — APPROVED** [review]
LGTM

File: infra/base-images/base-builder/jcc/jcc.go (hunk 3 above)
Comment (Alan32Liu) [reply infra/base-images/base-builder/jcc/jcc.go]: Yep, good point thanks!
Fixed : )

**Comment by Alan32Liu** [comment]
TBR

**Pull request merged**